flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/io/CheckpointBarrierUnalignerCancellationTest.java

```diff
@@ -0,0 +1,136 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.io;
+
+import org.apache.flink.runtime.checkpoint.CheckpointMetaData;
+import org.apache.flink.runtime.checkpoint.CheckpointMetrics;
+import org.apache.flink.runtime.checkpoint.CheckpointOptions;
+import org.apache.flink.runtime.checkpoint.channel.ChannelStateWriter;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.event.RuntimeEvent;
+import org.apache.flink.runtime.io.network.api.CancelCheckpointMarker;
+import org.apache.flink.runtime.io.network.api.CheckpointBarrier;
+import org.apache.flink.runtime.jobgraph.tasks.AbstractInvokable;
+import org.apache.flink.runtime.operators.testutils.DummyEnvironment;
+import org.apache.flink.util.function.RunnableWithException;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import javax.annotation.Nullable;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Future;
+
+import static org.junit.Assert.assertEquals;
+
+/**
+ * {@link CheckpointBarrierUnaligner} cancellation test.
+ */
+@RunWith(Parameterized.class)
+public class CheckpointBarrierUnalignerCancellationTest {
+	private final List<RuntimeEvent> events;
+	private final boolean shouldTriggerCheckpoint;
+	private final boolean shouldAbortCheckpoint;
+
+	public CheckpointBarrierUnalignerCancellationTest(boolean shouldTriggerCheckpoint, boolean shouldAbortCheckpoint, List<RuntimeEvent> events) {
+		this.events = events;
+		this.shouldTriggerCheckpoint = shouldTriggerCheckpoint;
+		this.shouldAbortCheckpoint = shouldAbortCheckpoint;
+	}
+
+	@Parameterized.Parameters(name = "should trigger: {0}, should abort {1}, events: {2}")
+	public static Object[][] parameters() {
+		return new Object[][]{
+				new Object[]{false, true, Arrays.asList(cancel(20), checkpoint(10))},
+				new Object[]{false, true, Arrays.asList(cancel(10), checkpoint(10))},
+				new Object[]{true, true, Arrays.asList(cancel(10), checkpoint(20))},
+				new Object[]{true, true, Arrays.asList(checkpoint(10), cancel(10))},
+				new Object[]{true, true, Arrays.asList(checkpoint(10), cancel(20))},
+		};
+	}
+
+	@Test
+	public void test() throws Exception {
+		TestInvokable invokable = new TestInvokable();
+		CheckpointBarrierUnaligner unaligner = new CheckpointBarrierUnaligner(new int[]{1}, ChannelStateWriter.NO_OP, "test", invokable);
+
+		for (RuntimeEvent e : events) {
+			if (e instanceof CancelCheckpointMarker) {
+				unaligner.processCancellationBarrier((CancelCheckpointMarker) e);
+			} else if (e instanceof CheckpointBarrier) {
+				unaligner.processBarrier((CheckpointBarrier) e, 0, 0);
```

Comment:
Should we also test `notifyBarrierReceived`?

Reply:
I don't think so because it is called from `processBarrier` (which is called here) and effects of it are essentially checked.